Structural/add custom scipy solver

manuelmessmer opened the pull request:
Provides stiffness and mass matrix as scipy matrices. 
This pull request shall resolve the issue #6111 opened by @RiccardoRossi . The implementation of the solver is pretty much what Riccardo suggested as a prototype.

manuelmessmer committed:
Adding custom scipy solver.

Provides stiffness and mass matrix as scipy matrices.

manuelmessmer committed:
Add custom_scipy to available solvers.

loumalouomega commented on kratos/python/add_strategies_to_python.cpp:
```diff
@@ -387,6 +387,7 @@ namespace Kratos
                 .def("BuildAndSolve", &BuilderAndSolverType::BuildAndSolve)
                 .def("BuildRHSAndSolve", &BuilderAndSolverType::BuildRHSAndSolve)
                 .def("ApplyDirichletConditions", &BuilderAndSolverType::ApplyDirichletConditions)
+                .def("ApplyConstraints", &BuilderAndSolverType::ApplyConstraints)
```
This changes should be in a independent PR..m the rest is OK

philbucher replied:
yes please make a PR for the changes in the Core

loumalouomega reviewed (dismissed):
Ok from my side

philbucher reviewed (changes requested):
Cool!
Blocking as I need some days to review

qaumann commented:
Thanks for the changes, I think this is very useful.
I am just generally wondering about the building process. Is the workaround using BUILD_LEVEL the way to go for this kind of problem or should a builder and solver provide the methods to build the different system matrices (e.g. BuildMassMatrix, BuildStiffnessMatrix, etc.)?

armingeiser commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py:
```diff
@@ -0,0 +1,147 @@
+    def __init__(self, main_model_part, custom_settings):
+        # Construct the base solver.
+        super(CustomScipySolver, self).__init__(main_model_part, custom_settings)
+        KratosMultiphysics.Logger.PrintInfo("::[CustomScipySolver]:: ", "Construction finished")
```
i would move this message to the bottom of the constructor

armingeiser commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py:
```diff
@@ -0,0 +1,147 @@
+        return solution_scheme
+
+    def _create_linear_solver(self):
+        ''' Linear solver will not be used. But eventually the solution stratgey calls the solver's clear function. 
```
```suggestion
        ''' Linear solver will not be used. But eventually the solution strategy calls the solver's clear function. 
```

armingeiser commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py:
```diff
@@ -0,0 +1,147 @@
+        
+    def SolveSolutionStep(self):
+        ## Obtain scipy matrices
+        # M = self._MassMatrixComputation()
```
why commented?

manuelmessmer replied:
Because that function should be overridden in the derived class. I just wanted to show how it works. Shall i uncomment it?

armingeiser replied:
If this function should always be overridden and never called directly, raise an error here
```python
        raise Exception("This function has to be implemented in the derived class")
```
The way it could be implemented could be explained in a docstring for this function, using the `"""` see e.g. kratos/python.

philbucher replied:
I would call it only if you have a dynamic scheme

RiccardoRossi replied:
this requires a special scheme...which is designed on purpose to compute the mass matrix and the stiffness matrix

armingeiser commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py:
```diff
@@ -0,0 +1,147 @@
+    def _create_mechanical_solution_strategy(self):
+        if self.settings["block_builder"].GetBool() == True:
+            warn_msg = "In case an eigenvalue problem is computed an elimantion builder shall be used to ensure boundary conditions are applied correctly!"
+            KratosMultiphysics.Logger.PrintWarning("CustomScipySolver", warn_msg)
```
Would an error work in this case? Or do you need the possibility to use a 'block_builder'

manuelmessmer replied:
Depends on what the user wants to do. I did it this way to be more general. In case user wants to have full matrix. If elimination builder only is enough error message would be better.

armingeiser replied:
Ok for me.

philbucher replied:
I don't understand why one cannot use the block-builder here

RiccardoRossi replied:
it is not strictly right that you cannot use it. the problem is that the eigenvalue associated to the diagonal which corresponds to the dirichlet modes(the "1s" in the diagonal is only controlled by the  value of the number you set in there. Depending on the chosen value it may interfere with the other eigenvalues

armingeiser reviewed (commented):
Looks good for me in general, i have only some small comments

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py:
```diff
@@ -0,0 +1,147 @@
+        The scheme determines the mass and stiffness matrices 
+        """
+        scheme_type = self.settings["scheme_type"].GetString()
+        if scheme_type == "dynamic":
```
static is missing

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py:
```diff
@@ -0,0 +1,147 @@
+    def _create_linear_solver(self):
+        ''' Linear solver will not be used. But eventually the solution stratgey calls the solver's clear function. 
+        To avoid crashing linear solver is provided here'''
+        return linear_solver_factory.CreateFastestAvailableDirectLinearSolver()
```
if you only need a dummy then this should return the baseclass

RiccardoRossi replied:
i agree

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_solver.py:
```diff
@@ -0,0 +1,147 @@
+def CreateSolver(main_model_part, custom_settings):
+    return CustomScipySolver(main_model_part, custom_settings)
+
+class CustomScipySolver(MechanicalSolver):
```
```suggestion
class MatrixBuilder(MechanicalSolver):
```

armingeiser replied:
:+1: or ScipyMatrixBuilder

RiccardoRossi replied:
Ouch i tend to disagree on this. this is a complete solver, which can be integrated together with the standard analysis stage...

armingeiser replied:
Now, with the implementation on SolveSolutionStep it would be something like "ScipyEigenvalueSolver"

RiccardoRossi replied:
true, nevertheless what is implemented there is just an example of what can be done... @manuelmessmer can u add a comment to tell that eigenvalue solution is only an example and that one should derive his own implementation from this "base solver"?

manuelmessmer replied:
Yes will do. I am also working on a test.
So I will keep the name "CustomScipySolver"?

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/python_solvers_wrapper_structural.py:
```diff
@@ -42,9 +42,12 @@ def CreateSolverByParameters(model, solver_settings, parallelism):
         else:
             err_msg =  "The requested solver type \"" + solver_type + "\" is not in the python solvers wrapper\n"
-            err_msg += "Available options are: \"static\", \"dynamic\", \"eigen_value\", \"harmonic_analysis\", \"formfinding\", \"adjoint_static\""
+            err_msg += "Available options are: \"static\", \"custom_scipy\", \"dynamic\", \"eigen_value\", \"harmonic_analysis\", \"formfinding\", \"adjoint_static\""
```
I don't think it should be here, 
Reason: this is quite sth special, that cannot be used without a corresponding `AnalysisStage`
Which btw I think you should add in this PR

RiccardoRossi replied:
why do you need a special analysis stage? I do think that one should be able to construct it...that would be very good for experiencing

philbucher replied:
my point is how do you use this solver?
It extends the interface of `PythonSolver` with the access-fct to the mass & stiffness matrix
=> this means that you need a custom analysis stage in any case, otherwise it does not make sense using this solver, bcs except building the matrices it does (and should not) do anything

RiccardoRossi replied:
as we spoke, this opens the door to experimenting with scipy. one should derive and use his experimental solver

manuelmessmer committed:
Adding function to assign eigenvectors to Kratos variable.

Riccardo asked me to add a function to be able to postprocess and visualize the eigenvectors.
Also made changes according to your suggestions.
Removed changes in the core. Will create new PR for that.

manuelmessmer commented:
> why do you compute this by hand?
> we have this functionality already in StructuralMechanics

Do you mean the assignment of the eigenvectors?

manuelmessmer committed:
Removed changes in kernel

philbucher commented:
no the entire computation of the eigenvalues.

philbucher commented:
https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py

philbucher commented:
@rishithellathmeethal could you please take a look at this? It is basically the same what we did in a hackish way some weeks ago

rishithellathmeethal commented:
> @rishithellathmeethal could you please take a look at this? It is basically the same what we did in a hackish way some weeks ago

ya, Its similar except that we wrote all these inside solvesolutionstep. 
I would like to have a function to get System vector too. Now I am accessing it from strategy using the "GetSystemVector"().

manuelmessmer commented:
> https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_eigensolver.py

Yes I know, but @RiccardoRossi wanted to solve the problem using scipy

philbucher commented:
Ok so given the discussion for me the usage/goal is still a bit unclear (@RiccardoRossi maybe you could answer since I think you came up with the design idea)
Is this solver supposed to be used for 
    a) computing the stiffness (& mass) -matrix or 
    b) for solving the eigenvalue problem with scipy?
I think a) by itself is already very useful and would help many ppl. This howerver is a special usecase that cannot be integrated in the regular workflow, bcs it needs a custom `AnalysisStage` for calling the matrix-access-fcts

b) Adding b) I think is also fine even though I don't see the usecase as clear. I guess using scipy instead of our own eigenvalue-solver would be nice to have. (implementational detail: instead of reimplementing `_AssignVariables` the C++-version already existing in the `EigenSolverStrategy` should be used, which might need some refactor)

I think what could make sense is the following desing:
~~~py
class ScipyMatrixBuilder(MechanialSolver):
    def ComputeStiffnessMatrix(self):
        # compute stiffness matrix, to be called from a custom analysis_stage
        return stiffness_matrix

    def ComputeMassMatrix(self):
        # compute mass matrix, to be called from a custom analysis_stage
        return mass_matrix
  
class ScipyEigenSolver(ScipyMatrixBuilder):
    def SolveSolutionStep(self):
        K = self.ComputeStiffnessMatrix()
        M = self.ComputeMassMatrix()
        # solve eigenvalue problem with scipy
~~~

what do you think?

RiccardoRossi commented:
Hi @philbucher i designed this as a "base" solver.
my idea is that someone can hack this soler by **creating a derived one** in which one can prototype, in python, what he wants.

the first usecase could be to compute eigenvalues, but the scope is much larger than that.

Essentially i am giving access to (a copy of) K,M (and as a second step i would like to add D) as well as to the doflist. 
one can do what he likes with those (in his own derived solver) and mock up in python a complete solver.

this is a very useful think to protype stuff at the python level, and definitely to teach for example about the application of constraints.

I am not in favor of splitting it as you propose, since the way it is done now the initialization and all the rest is done there where it should be.

philbucher commented:
@RiccardoRossi ok this makes sense to me
Only thing is that IMO the computation of Eigenvalues should not be done in a baseclass, otherwise I agree with you

BTW this implies that we should finally make the usage of custom solvers easier => I have a solution for the same problem in the CoSimApp, I will try in the next days to port it also to Structural

philbucher commented:
So I just talked to @RiccardoRossi and want to share what we discussed:
- the `CustomScipySolver` is to be considered as a baseclass for solvers with which "hacking" can be done in python. Hence it could be clearer to rename it to `CustomScipyBaseSolver` or sth similar
- the computation of the eigenvalues is an EXAMPLE to show a usecase (@manuelmessmer please leave then the AssignVariables` in python and put some comments making it clear that this is only an example for the usage
- I will implement the feature to construct custom solvers, then we don't need to add every solver by hand into the wrapper. I have this already implemented in CoSim, I only have to port this functionality

What I think is missing now is cleanup and I suggest also a test (using the eigenvalue computation)

philbucher commented:
> I will implement the feature to construct custom solvers, then we don't need to add every solver by hand into the wrapper. I have this already implemented in CoSim, I only have to port this functionality

see #6250

manuelmessmer committed:
Renamed Solver and made some comments.

manuelmessmer committed:
Removed Solver from StructuralSolverWrapper.

This way it will be used: #6250

manuelmessmer committed:
Added test case

Test computes eigenvalue problem and compares scipy solution with standard eigen solution.
Scipy solver is constructed to suit #6250.

manuelmessmer commented on applications/StructuralMechanicsApplication/tests/test_custom_scipy_base_solver.py:
```diff
@@ -0,0 +1,149 @@
+                "end_time"      : 1.0
+            },
+            "solver_settings" : {
+                "solver_type"              : "structural_mechanics_custom_scipy_base_solver",
```
Please note: Test will fail. As I just provide the name of the solver to suit #6250.

RiccardoRossi commented:
Hi @philbucher 
Are we waiting for something to merge this?

philbucher commented:
Your review on #6250

philbucher commented:
@manuelmessmer since #6250 is merged can you please merge master here such that we can go ahead with this PR?

manuelmessmer committed:
Merge branch 'master' into structural/add_custom_scipy_solver

manuelmessmer commented:
Do we need anything more here?

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py:
```diff
@@ -0,0 +1,198 @@
+        return KratosMultiphysics.LinearSolver()
+
+    def _create_mechanical_solution_strategy(self):
+        if self.settings["block_builder"].GetBool() == True:
```
```suggestion
        if self.settings["block_builder"].GetBool():
```

more pythonic

philbucher commented on applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_custom_scipy_base_solver.py:
```diff
@@ -0,0 +1,198 @@
+    def __init__(self, main_model_part, custom_settings):
+        # Construct the base solver.
+        super(CustomScipyBaseSolver, self).__init__(main_model_part, custom_settings)
+        self.space = KratosMultiphysics.UblasSparseSpace()
```
Any particular reason this is a member?
Unless you have one, I would remove it and use it locally
It is the baseclass, derived classes might not need it

manuelmessmer replied:
Well it is used in _MassMatrixComputation and _StiffnessMatrixComputation, so probably also the derived classes will need it, correct?
But I can also declare it as a local variable in both methods?

philbucher replied:
I would make them local variables
Note that the methods in the Space are all `static` hence I think this makes sense

manuelmessmer replied:
Ok will do!

philbucher commented on applications/StructuralMechanicsApplication/tests/test_custom_scipy_base_solver.py:
```diff
@@ -0,0 +1,149 @@
+    props[StructuralMechanicsApplication.I33] = 0.00001
+    props[StructuralMechanicsApplication.USE_CONSISTENT_MASS_MATRIX] = True
+
+    for i_elem, connectivity in enumerate(element_connectivities):
```
+1 for using `enumerate`

philbucher commented on applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py:
```diff
@@ -306,6 +308,8 @@ def AssembleTestSuites():
     smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TTestPostprocessEigenvaluesProcess]))
     smallSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TestEigenSolverWithConstraints]))
+    # Custom Scipy Solver test
+    nightSuite.addTests(KratosUnittest.TestLoader().loadTestsFromTestCases([TestCustomScipyBaseSolver]))
```
how long does this test take on your machine?

manuelmessmer replied:
I believe to remember it was relatively long, although it is pretty much the same example as in the normal EigensolverTest.  I let you know, but first have to recompile.

manuelmessmer replied:
It takes 0.5sec

philbucher replied:
ok then nightly is appropriate

philbucher reviewed (commented):
minor comment otherwise I am positive about this PR!

philbucher commented:
@manuelmessmer check codacy, it has some suggestions

Esp I would set the editor to auto delete whitespaces when saving a file. With VSCode this is very easy

manuelmessmer committed:
Review comments.

philbucher reviewed (approved):
Approving since this has been there for a long time and I think everyone had time to raise their opinion.
Nice work and thanks for the patience @manuelmessmer

philbucher commented:
@manuelmessmer the test fails if scipy is not available
can you please add a `skiptest`?

manuelmessmer commented:
I pushed the changes here #6791.
